# ml_zymoseptoria

## Fungal Pathogen Gene Selection for Predicting the Onset of Infection Using a Multi-Stage Machine Learning Approach

*Authors Oliver Stoner, Graham Thomas, Fabrizio Costa, Ryan M Ames*

Phytopathogenic fungi pose a serious threat to global food security. Next-generation sequencing technologies, such as transcriptomics, are increasingly used to profile infection, assess environmental adaptation and gauge host-responses. The accumulation of these large-scale data has created the opportunity to employ new computational methods to gain greater biological insights. Machine learning approaches that learn to identify patterns in complex datasets have recently been applied to the field of plant-pathogen interactions. Here, we apply a machine learning approach to transcriptomics data for the fungal pathogen Zymoseptoria tritici to predict the onset of infection as measured by disease symptom development. We show that we can identify the most important genes that predict infection timings, accurately classify strains as early and late infectors and predict the timing of infection of ‘novel’ strains using only a subset of the data. These methods and the genes identified further demonstrate the use of these methods in the field of plant-pathogen interactions and have implications for the identification of biomarkers for disease monitoring and forecasting.

An R Notebook containing all code and visualisations from this work can be viewed [here](ML_Zymo.nb.html) and downloaded [here](ML_Zymo.Rmd)
